hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanner.java

```diff
@@ -370,6 +376,59 @@ private String getLatestVersionBeforeCommit(List<FileSlice> fileSliceList, Hoodi
     return earliestCommitToRetain;
   }
 
+  public Map<String, List<String>> getReplacedFileIdsToClean(Option<HoodieInstant> earliestInstantToRetain) {
```

Comment:
fileSystemView#getReplacedFileGroupsBeforeOrOn looks similar (we may have to add another method 'getReplacedFileGroupsBefore' to enforce strict inequality). Maybe we can reuse code?  I can do this later on if we file code cleanup task.